vault/cluster.go

```diff
@@ -321,6 +321,11 @@ func (c *Core) startClusterListener(ctx context.Context) error {
 		// If we listened on port 0, record the port the OS gave us.
 		c.clusterAddr.Store(fmt.Sprintf("https://%s", c.getClusterListener().Addr()))
 	}
+
+	if err := c.getClusterListener().SetAdvertiseAddr(c.ClusterAddr()); err != nil {
+		return err
+	}
+
```

Comment:
I'm a bit confused on how this works. `c.getClusterListener().Addr()` returns back `cl.Addrs()[0]` the first time around, so could we simply do the parsing of the address into an absolute one and return that as part of `Addr()` instead of adding a new field and setter just for this?

Reply:
Yeah this is primarily necessary because our tests use random port assignments, so we actually don't know the port until we start the listener. If we didn't do this here then each test node would be advertising `127.0.0.1:0`

Reply:
I see. My only concern here is that the logic right above could be overwriting `c.clusterAddr` to be a value a non-specified IP address value.

The call to c.getClusterListener().Addr()) could get us back an address from the listener config that's `0.0.0.0:8201` for instance, so we'd be updating `c.clusterAddr` to be `https://0.0.0.0:8201`.

Reply:
At least raft will then reject the IP. The user will need to re-configure it such that one of the two values is a fully formed hostname.

Reply:
True. I was trying to think of other potential alternatives to this logic, since  `c.clusterAddr.Store(fmt.Sprintf("https://%s", c.getClusterListener().Addr()))` followed by `c.getClusterListener().SetAdvertiseAddr(c.ClusterAddr())` felt a little weird. IIUC it's taking the listener's address to update the cluster address and then the listener's address is updated with this cluster address, but it all seemed a little loopy.

Reply:
It is a little indirect, but i don't see a great way around that at the moment. Potentially we can push the `:0` port check into test code since that's where it's needed.